pkg/validation/variables/variables_component_test.go

```diff
@@ -0,0 +1,212 @@
+package variables
+
+import (
+	"testing"
+
+	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestValidateAndReplaceContainerComponent(t *testing.T) {
+
+	tests := []struct {
+		name         string
+		testFile     string
+		outputFile   string
+		variableFile string
+		wantErr      bool
+	}{
+		{
+			name:         "Good Substitution",
+			testFile:     "test-fixtures/components/container.yaml",
+			outputFile:   "test-fixtures/components/container-output.yaml",
+			variableFile: "test-fixtures/variables/variables-referenced.yaml",
+			wantErr:      false,
+		},
+		{
+			name:         "Invalid Reference",
+			testFile:     "test-fixtures/components/container.yaml",
+			outputFile:   "test-fixtures/components/container.yaml",
+			variableFile: "test-fixtures/variables/variables-notreferenced.yaml",
+			wantErr:      true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			testContainerComponent := v1alpha2.ContainerComponent{}
+			readFileToStruct(t, tt.testFile, &testContainerComponent)
+
+			testVariable := make(map[string]string)
+			readFileToStruct(t, tt.variableFile, &testVariable)
+
+			err := validateAndReplaceForContainerComponent(testVariable, &testContainerComponent)
+			verr, ok := err.(*InvalidKeysError)
+			if tt.wantErr && !ok {
+				t.Errorf("Expected InvalidKeysError error from test but got %+v", verr)
```

Comment:
Same comments as for `variables_command_test.go`

Reply:
updated to `err` for all the tests